### PR TITLE
Make Real Audio files don't parse with MediaInfo

### PR DESCRIPTION
**Pull request by Sami32**

### CODE REVIEW

**Comment by Nadahar** [comment]
@Sami32 I haven't had a problem parsing RA files with MediaInfo:

```logtalk
Parsing results for file "sample.ra": Container: RM, Size: 28431, Bitrate: 64000, Duration: 00:00:03.00, Audio Tracks: 1 [Audio Codec: AC3, Bitrate: 64000, Channel(s): 1, Sample Frequency: 11025 Hz], Mime Type: audio/mpeg
```
```logtalk
General
Complete name                            : E:\data\Sourcecode\Java\Media Sample Audio\ra\sample.ra
Format                                   : RealMedia
File size                                : 27.8 KiB
Duration                                 : 3s 482ms
Overall bit rate                         : 64.0 Kbps

Audio
ID                                       : 0
Format                                   : AC-3
Format/Info                              : Audio Coding 3
Codec ID                                 : dnet
Codec ID/Info                            : Real Player 3
Duration                                 : 3s 482ms
Bit rate                                 : 64.0 Kbps
Channel(s)                               : 1 channel
Sampling rate                            : 11.025 KHz
Bit depth                                : 16 bits
Compression mode                         : Lossy
Stream size                              : 27.2 KiB (98%)
```

```logtalk
Parsing results for file "HugeWAV3_arch001.ra": Container: RM, Size: 152192, Bitrate: 64082, Duration: 00:00:17.00, Audio Tracks: 1 [Audio Codec: Cook, Bitrate: 64082, Channel(s): 1, Sample Frequency: 44100 Hz], Mime Type: audio/mpeg
```
```logtalk
General
Complete name                            : E:\data\Sourcecode\Java\Media Sample Audio\ra\HugeWAV3_arch001.ra
Format                                   : RealMedia
File size                                : 149 KiB
Duration                                 : 17s 660ms
Overall bit rate                         : 64.1 Kbps

Audio
ID                                       : 0
Format                                   : Cooker
Codec ID                                 : cook
Codec ID/Info                            : Based on G.722.1, Real Player 6
Duration                                 : 18s 558ms
Bit rate                                 : 64.1 Kbps
Channel(s)                               : 1 channel
Sampling rate                            : 44.1 KHz
Bit depth                                : 16 bits
Stream size                              : 145 KiB (98%)
```

**Comment by Nadahar** [comment]
Also, as I said in the discussion about DSD and ADPCM, UMS **doesn't** parse audio files with FFmpeg. It seems it never did either: 079805f534da336ef84f4c5ba7ba998a8a1705b6.

**Comment by Sami32** [comment]
Strange...could you send 1 or 2 samples ? I suspect your samples to be RM or RMVB renamed to RA.
Here are mine:
[ra4-28_8.ra.txt](https://github.com/UniversalMediaServer/UniversalMediaServer/files/900866/ra4-28_8.ra.txt)
[ddinterview28.ra.txt](https://github.com/UniversalMediaServer/UniversalMediaServer/files/900870/ddinterview28.ra.txt)

Thank you for sharing these informations.

**Comment by Nadahar** [comment]
@Sami32 You are correct, your files can't be parsed using MediaInfo. However, it seems that ```JAudioTagger``` can:

```logtalk
Audio Header content:
	bitRate:64
	isVbr:false
	trackDuration:17.0

REAL Tag content:
	TITLE:
	ARTIST:
	COMMENT:
-------------------
```

**Comment by Sami32** [comment]
Perhaps it only affect the lossless ones ?

**Comment by Sami32** [comment]
_JAudioTagger_ results are very different from _FFmpeg_ ones...

**Comment by Nadahar** [comment]
@Sami32 No, ```JAudioTagger``` is what's used when MediaInfo isn't - so you solution is valid in that sense (except that you claim it to be FFmpeg).

There are very few metadata available for this one file though, and there's no "detection" of the format or audio codec. I just looked in ```JAudioTagger```'s code, and they simply use extensions ```ra``` and ```rm``` for recognition - so we might as well use ```Format``` (which is based on extension) to decide just like I did with DSD and ADPCM. The thing is that parsing based on extension isn't something I really like doing, it's not a very "safe" way of doing it.

**Comment by Nadahar** [comment]
@Sami32 According to [this]() there are "two types" of real audio files:

RA 	  	RealAudio file:
```
2E 52 4D 46 00 00 00 12 00 	  	.RMF.....
```
RA 	  	RealAudio streaming media file:
```
2E 72 61 FD 00 	  	.raý.
```

I've checked with a hexeditor, and the two samples I had (which are parsed by MediaInfo) have the ```RMF``` signature, while your samples have the ```raý``` signature. That makes it a bit complicated...

**Comment by Sami32** [comment]
No ? I see that you only tested the first file. You don't want send me a sample ? It's for my test library ;)
Bitrate ;
JAudioTagger ==> 64    (Duration 17 mn)
FFmpeg ==> 15        (Duration 1.44)
Potplayer ==> 7 to 20

**Comment by Sami32** [comment]
Yes, it what i say:
You have `cook` and i have `ralf`, i guess.

**Comment by Sami32** [comment]
It's seem that you have RM file renamed to RA.

**Comment by Nadahar** [comment]
@Sami32 I don't think they are "renamed" files, I think there are two formats using the ```ra``` extension. If you see the information I linked above, it supports that. According to that information, "your" files are "	RealAudio streaming media file" - and that's the type that MediaInfo can't parse.

The wildly varying parsing is also interesting... what is the actual playlength?

**Comment by Sami32** [comment]
In fact it seem that you files can have any of the 2 extensions:
2E 52 4D 46 	  	.RMF
RM, RMVB 	  	RealMedia streaming media file

2E 52 4D 46 00 00 00 12
00 	  	.RMF....
.
RA 	  	RealAudio file

2E 72 61 FD 00 	  	.raý.
RA 	  	RealAudio streaming media file

**Comment by Nadahar** [comment]
[This](https://www.velocityreviews.com/threads/realaudio-file-header.401453/) also states that a RA file has the RMF signature.

**Comment by Sami32** [comment]
So why not make it simple:
When the format is undertiminated defer to an other parser than MediaIfo ?

**Comment by Nadahar** [comment]
> When the format is undertiminated defer to an other parser than MediaIfo ?

That would be a good general approach, but there are some details that must be figured out. For example, what information must be missing for it to be declared "failed" by MediaInfo. Also, the ```JAudioTagger``` results seems to be lacking, and we have no way except file extension to "classify" the files.

**Comment by Sami32** [comment]
We have crossposted. The real duration is around 1s so FFmeg is the most accurate with 1.44

For MediaInfo i think that unknown container should be the safe way to go, that will need to come after your DSDAudio and ADPCM categories test.

About _JAudioTagger,_ i have already submitted this and many other issues, so it could be fixed in the next 2.2.6 version...or not (some are already)
_FFmpeg_ is able to identify them as `ra_288`.

So for now, i'm not sure how to handle it the more correctly.

**Comment by Nadahar** [comment]
RA is really a mess of secrets, trickery, proprietary licenses and general "cleverness" to try to cheat everyone as much as possible. It seems there are many versions of RA files, and they can carry a variety of codecs. I don't know if it's worth trying to figure this out. RealMedia's so called "open source" project Helix is cancelled from what I can see, and not much is actually open source. Most relevant information has been removed from the net, here are some more or less relevant links:
https://web.archive.org/web/20070617085736/http://service.real.com/help/library/guides/production/htmfiles/audio.htm#12311
http://homerecording.com/realaudio.html

**Comment by Nadahar** [comment]
Actually, the [last link above](https://wiki.multimedia.cx/index.php/RealMedia#RealAudio_1.0_file_.28.ra_version_3.29) turned out to be helpful to figure this mess out. It's the early (old) versions that use the ```.raý``` signature. They have very little metadata to begin with, so there's not much to parse, especially version 3. Both "your" files were version 4 of RA (but RealAudio version 2 😕 ), they start with:
2E72 61FD 0004 (```.r``` ```a²``` ```null EOT```). The last byte is the version number, from 3 to 5. 

These files are always CBR, so that part can be deducted from detecting the signature bytes. In version 3 there's really nothing more that the title to parse. In version 4 there are some parsable fields, but no length. I guess that explains why the lenghts are different, they probably try to calculate it based on the header information and the size of the data. 
```
word     Codec flavor
dword    Coded frame size
byte[12] Unknown
word     Sub packet h
word     Frame size
word     Subpacket size
word     Unknown
word     Samplerate
word     Unknown
word     Sample size
word     Channels
byte     Interleaver ID string length (always 4)
byte[]   Interleaver ID string
byte     FourCC string length (always 4)
byte[]   FourCC string
byte[3]  Unknown
byte     Title string length
byte[]   Title string
```

**Comment by Sami32** [comment]
Good link :+1: 
I used _webarchive_ of Real database as well, while searching their official/correct MIME types to be used. But as you said it's difficult to find as nothing more is kept in live on their website.
_Helix_ is used in _FFmpeg_.

That's said, i don't have any other samples to test, and i don't know how you do test using _JAudioTagger_ ? With a specific software or Eclipse ?

I don't know for your files but perhaps _JAudioTagger_ give more accurate result on them, in thet case we don't necessary need to bother.
The second file thet i sended you have more metadatas in it.

Is that you files will be fine being parsed with _JAudioTagger_ ? That'll make thing simpler.

**Comment by Nadahar** [comment]
@Sami32 I simply debug with Eclipse. Our code is messy enough that even if MediaInfo is used, it still goes through the ```JAudioTagger``` parsing when doing the "thumbsonly" run. Since it's "thumbsonly" no results are kept, but using Eclipse's debugger I can see all the values being parsed.

I'll do a test-parse on your files and see...

**Comment by Sami32** [comment]
Thanks, as without your samples or using Eclipse i'm stuck.

Though if your samples are correctly parsed with _JAudioTagger_ i'll consider this PR as done, as it'll be an enhancement.

**Comment by Nadahar** [comment]
@Sami32 Sorry, "my" files parse with JAudioTagger AND MediaInfo. "Yours" on the other hand, the old ones, don't:
```
org.jaudiotagger.audio.exceptions.CannotReadException: Corrupt file: RealAudio chunk length at position 262148 cannot be less than 8
```

**Comment by Sami32** [comment]
Never mind. So since you didn't elaborate, i deduct that MediaInfo and JAudioTagger have the same or very similar parsing.

I guess that when you say "old ones" you referred to _ra4-28_8.ra_ ? because since you say that one of my file was parsed by _JAudioTagger_, i just become lost...
It's why i asked you at first, before make that PR, as i was not sure if my files was corrupt, as one have a very little size but they all play fine.

**Comment by Nadahar** [comment]
@Sami32 I didn't elaborate because I've pasted MediaInfo's parsing of the "newer" files above, and it looks perfectly fine. What I call the "older" files is RealAudio 1.0 and RealAudio 2.0 - which were used before they created the RealMedia format which handles both audio and video. RealMedia can still be audio files, but it's a completely different format. 

Neither JAudioTagger nor MediaInfo parses version 1.0/2.0 at all.

I'm sorry that I didn't check it before you started the PR, but there are constantly so many things I should do that I loose track.

**Comment by Sami32** [comment]
Don't worry, i forget as well ;)


What you try to mean is that this PR didn't improve the RA support, is that correct ?

**Comment by Nadahar** [comment]
@Sami32 Not as it is, but hang on - I'm trying to put something together that you can work further on...

**Comment by Sami32** [comment]
That's nice but don't bother you with that, as it's a deprecated format, you work already on too much things.

I don't really see what you mean by "something", as except using _FFmpeg_ ...

**Comment by Nadahar** [comment]
@Sami32 Here's [something](https://github.com/UniversalMediaServer/UniversalMediaServer/tree/ParseRealAudio).

It sort of works, but there are some things I haven't figured out. I added a lot of traces in there so that you can try to debug it further. First of all, the test files I have is "two bytes out of sync" with [this description](https://wiki.multimedia.cx/index.php/RealMedia#RealAudio_2.0_file_.28.ra_version_4.29) (I've had to disable one of the retrieved fields, or everything was parsed wrong - the question is why and if there are different versions of this format with different length, or if the info on that site is simply wrong). 

Second, you have to look at the sample frequency and bit rate stuff - I'm not sure what to put in those values - and last - we have to figure out how to calculate the duration. You might find some existing code to take a look in..?

I have to go do something else now so I really can't continue now, but I'm hoping that you can solve some of this 😉

**Comment by Sami32** [comment]
Excellent work, as always :wink: 

I didn't intended to go that deep for this desuet format, it's take me the fun out ;)

**Comment by Nadahar** [comment]
@Sami32 You don't have to close this, there's no reason to make yet another PR for this. I can just "move" the commit over to this branch when we get it working properly.

**Comment by Nadahar** [comment]
@Sami32 Have you looked at the things missing in the parsing?

**Comment by Nadahar** [comment]
It's not just I who call version 1.0 and 2.0 for the "old" format, FFmpeg code:
```c
    if (tag == MKTAG('.', 'r', 'a', 0xfd)) {
        /* very old .ra format */
        return rm_read_header_old(s);
    } else if (tag != MKTAG('.', 'R', 'M', 'F')) {
        return AVERROR(EIO);
    }
```

**Comment by Sami32** [comment]
No, i didn't. Because i was not going to go that deep and don't have motivation or interest to it more than the PR that i have push.
Sorry, but my fun stopped when i saw too many lines of code, for some that i don't even understand, for a simple RA improvement idea that i was naively thought easy...

**Comment by Nadahar** [comment]
@Sami32 Ok.. I have to remember to stay away from things, it's a shame that I made this for nothing. There are only some minor details missing, but it isn't very useful without figuring out how to get these details right.

**Comment by Sami32** [comment]
@Nadahar Sorry :disappointed:  i didn't mean to offense you. At first i just asked for confirmation, if that will be fine to do and if that'll work.

I didn't expect that you go so far on it ! Sorry for not being clear and make you in trouble.
That doesn't mean that you great work will not be used.

**Comment by Nadahar** [comment]
@Sami32 I know you didn't ask me to do that, but I saw that we had no way of parsing the "old" RealAudio files and I thought that it would be easy to parse such simple files. 

The problem is that the information I used (from the link above) doesn't seem to be 100% accurate, so unless someone figures out what's wrong with the format description is won't be used. It doesn't parse correcly as it is now.

**Comment by Sami32** [comment]
I really don't know if it can be of any use to you, but just in case:
https://github.com/FFmpeg/FFmpeg/blob/415f907ce8dcca87c9e7cfdc954b92df399d3d80/libavcodec/ra288.h

I only read that RA 1.0 version was 1 channel with a fixed bitrate of 8000, and the 2.0 version with 1 or 2 channels and a fixed bitrate of 8000.

**Comment by Nadahar** [comment]
@Sami32 I've already looked in the FFmpeg code quite a bit (not ra288.h though as that's a codec). I don't think I've found the correct part yet though, I really wish I had a proper C IDE configured as there are so many references that's very hard to follow "manually".
